backbone/src/main/java/org/researchstack/backbone/ui/step/layout/FingerprintStepLayout.java

```diff
@@ -0,0 +1,479 @@
+                            handleFingerprintSuccess();
+                        }
+                    }, getResources().getInteger(R.integer.rsb_config_anim_time_check_mark));
+                }
+            });
+        }
+
+        @Override
+        public void onAuthenticationFailed() {
+            // Callback is not on the main thread, so send it to the main thread
+            removeCallbacks(fingerprintAnimationRunnable);
+            post(new Runnable() {
+                @Override
+                public void run() {
+                    showError(getContext().getResources().getString(R.string.rsb_fingerprint_not_recognized));
+                }
+            });
+        }
+    };
+
+    protected void handleFingerprintSuccess() {
+        if (fingerprintStep.isCreationStep()) {
+            generateAndEncryptPin();
+        } else {
+            injectPin();
+        }
+        stopListening();
+        callbacks.onSaveStep(StepCallbacks.ACTION_NEXT, fingerprintStep, null);
+    }
+
+    /**
+     * Creates a symmetric key in the Android Key Store, which can only be used after the user has
+     * authenticated with fingerprint.
+     */
+    private void createKey() {
+        // The enrolling flow for fingerprint. This is where you ask the user to set up fingerprint
+        // for your flow. Use of keys is necessary if you need to know if the set of
+        // enrolled fingerprints has changed.
+        try {
+            KeyGenerator keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE);
+            keyGenerator.init(
+                    new KeyGenParameterSpec.Builder(KEY_NAME,
+                            KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
+                            .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+                            // we know our pincode is securely random, since we generate it ourselves,
+                            // so we do not need an Initialization Vector (IV)
+                            .setRandomizedEncryptionRequired(false)
+                            .setUserAuthenticationRequired(true)
+                            .build());
+            keyGenerator.generateKey();
+
+        } catch (NoSuchAlgorithmException | InvalidAlgorithmParameterException | NoSuchProviderException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Initialize the {@link Cipher} instance with the created key in the {@link #createKey()}
+     * method.
+     *
+     * @return {@code true} if initialization is successful, {@code false} if the lock screen has
+     * been disabled or reset after the key was generated, or if a fingerprint got enrolled after
+     * the key was generated, which in all cases we must ask the user to sign back in with their DataProvider credentials
+     */
+    private boolean initCipher() {
```

Comment:
No, you are right. The keystore/fingerprint/encryption code should go in some sort of security utils class. I'll do that.